src/mesh/domain.hpp

```diff
@@ -114,68 +143,124 @@ class IndexShape {
+  // is sufficient to fill whole ghost layer
   KOKKOS_INLINE_FUNCTION int ks(const IndexDomain &domain) const noexcept {
-    return (domain == IndexDomain::entire) ? 0 : x_[2].s;
+    switch (domain) {
+    case IndexDomain::entire:
+      return 0;
+    case IndexDomain::inner_x3:
+      return 0;
+    case IndexDomain::outer_x3:
+      return entire_ncells_[2] == 1 ? 0 : x_[2].e + 1;
+    default: // interior+
+      return x_[2].s;
+    }
   }
 
   KOKKOS_INLINE_FUNCTION int ie(const IndexDomain &domain) const noexcept {
-    return (domain == IndexDomain::entire) ? entire_ncells_[0] - 1 : x_[0].e;
+    switch (domain) {
+    case IndexDomain::interior:
+      return x_[0].e;
+    case IndexDomain::inner_x1:
+      return x_[0].s == 0 ? 0 : x_[0].s - 1;
+    default:
+      return entire_ncells_[0] - 1;
+    }
   }
 
   KOKKOS_INLINE_FUNCTION int je(const IndexDomain &domain) const noexcept {
-    return (domain == IndexDomain::entire) ? entire_ncells_[1] - 1 : x_[1].e;
+    switch (domain) {
+    case IndexDomain::interior:
+      return x_[1].e;
+    case IndexDomain::inner_x2:
+      return x_[1].s == 0 ? 0 : x_[1].s - 1;
+    default:
+      return entire_ncells_[1] - 1;
+    }
   }
 
   KOKKOS_INLINE_FUNCTION int ke(const IndexDomain &domain) const noexcept {
-    return (domain == IndexDomain::entire) ? entire_ncells_[2] - 1 : x_[2].e;
+    switch (domain) {
+    case IndexDomain::entire:
+      return entire_ncells_[2] - 1;
+    case IndexDomain::inner_x3:
+      return x_[2].s == 0 ? 0 : x_[2].s - 1;
+    case IndexDomain::outer_x3:
+      return entire_ncells_[2] - 1;
+    default: // interior+
+      return x_[2].e;
+    }
   }
 
   KOKKOS_INLINE_FUNCTION int ncellsi(const IndexDomain &domain) const noexcept {
-    return (domain == IndexDomain::entire) ? entire_ncells_[0] : x_[0].e - x_[0].s + 1;
+    if (entire_ncells_[0] == 1 &&
+        (domain == IndexDomain::inner_x1 || domain == IndexDomain::outer_x1)) {
+      return 0; // if x1 is zero-dimensional, there are no ghost zones
+    }
+    auto bnds = GetBoundsI(domain);
+    return bnds.e - bnds.s + 1;
   }
 
   KOKKOS_INLINE_FUNCTION int ncellsj(const IndexDomain &domain) const noexcept {
-    return (domain == IndexDomain::entire) ? entire_ncells_[1] : x_[1].e - x_[1].s + 1;
+    if (entire_ncells_[1] == 1 &&
+        (domain == IndexDomain::inner_x2 || domain == IndexDomain::outer_x2)) {
+      return 0; // if x2 is zero-dimensional, there are no ghost zones
+    }
+    auto bnds = GetBoundsJ(domain);
+    return bnds.e - bnds.s + 1;
   }
 
   KOKKOS_INLINE_FUNCTION int ncellsk(const IndexDomain &domain) const noexcept {
-    return (domain == IndexDomain::entire) ? entire_ncells_[2] : x_[2].e - x_[2].s + 1;
+    if (entire_ncells_[2] == 1 &&
+        (domain == IndexDomain::inner_x3 || domain == IndexDomain::outer_x3)) {
+      return 0; // if x3 is zero-dimensional, there are no ghost zones
+    }
+    auto bnds = GetBoundsK(domain);
+    return bnds.e - bnds.s + 1;
   }
```

Comment:
Can you make bnds const auto &?

Reply:
See my above comment. `GetBounds*` returns a whole new `IndexRange object. I can make it `const` but not `const auto &`. We could maybe pre-compute and store the `Bounds` objects... but that seems like it would require some surgery.

Reply:
If not a method should probably be added that does this.

Reply:
Ah, ya you should just use the `ncells` methods.

Reply:
Oh I see. Yes, that's better.